lib/classes/Variables.test.js

```diff
@@ -262,6 +262,47 @@ describe('Variables', () => {
       });
     });
 
+    it('should NOT call populateObject if variable value is a circular object', () => {
+      const serverless = new Serverless();
+
+      serverless.variables.options = {
+        stage: 'prod',
+      };
+      const property = '${opt:stage}';
+      const variableValue = {
+        stage: '${opt:stage}',
+      };
+      const variableValuePopulated = {
+        stage: 'prod',
+      };
+
+      serverless.variables.populatedObjectReferences['opt:stage'] = variableValuePopulated;
+
+      serverless.variables.loadVariableSyntax();
+
+      const populateObjectStub = sinon
+        .stub(serverless.variables, 'populateObject')
+        .resolves(variableValuePopulated);
+      const getValueFromSourceStub = sinon
+        .stub(serverless.variables, 'getValueFromSource')
+        .resolves(variableValue);
+      const populateVariableStub = sinon
+        .stub(serverless.variables, 'populateVariable')
+        .resolves(variableValuePopulated);
+
+      return serverless.variables.populateProperty(property).then(newProperty => {
+        expect(populateObjectStub.called).to.equal(false);
+        expect(getValueFromSourceStub.called).to.equal(true);
+        expect(populateVariableStub.called).to.equal(true);
+        expect(newProperty).to.deep.equal(variableValuePopulated);
+
+        serverless.variables.populateObject.restore();
```

Comment:
Cleanup should be moved in an `afterEach` block so that the `restore` is still done even if this test fails.